woocommerce/woocommerce-gutenberg-products-block
# Introduce ExtendRestApi

Part of https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/3244

This PR introduces `ExtendRestApi` a class that 3PD should use to extend the StoreAPI response to add extra data.
The goal of using this instead of filters directly is:
- Protecting the integrity of the original checkout, so that people don't break existing flows.
- Protecting other 3PD from having their code being broken by a rogue plugin.
- Protecting core experience from breaking if a plugin malfunction.
- Ensuring any integration is well validated and has the correct data.

For this, we changed how AbstractSchema work to be aware of `ExtendRestApi`.

### Interface.
3PD should call a `register_endpoint_data` function, that adds correct schema and data to each endpoint.

### Signature 

```php
$ExtendRestApi->register_endpoint_data( string $namespace, string $endpoint, callable $schema_callback, callable $data_callback );
```

$schema_callback is a function that should return an array with metadata of new endpoint being registered:

```php
function mySchemaFunction() {
  return [
		'fee' => [
			'description' => __( 'X Product fees', 'my-plugin' ),
			'type'        => 'integer',
			'context'     => [ 'view', 'edit' ],
			'readonly'    => true,
		],
	];
}
```

$data_callback is a function that should return an array with actual data being registered:

```php
function myDataFunction( $cart_item ) {
	$data = []
	if ( $cart_item['type'] === 'x-product' ) {
		$myXProduct = new MyCustomProduct( $cart_item['item'] );
		$data = [
			'fee' => $myXProduct->get_fee();
		]
	}
  return $data;
} 
```

### Current endpoints that can be extended.
Currently, we only have `cart/items` that can be extended, it passes `$cart_item` to `$data_callback`.

### Code example

```php
use Automattic\WooCommerce\Blocks\Package;
use Automattic\WooCommerce\Blocks\Domain\Services\ExtendRestApi;
use Automattic\WooCommerce\Blocks\StoreApi\Schemas\CartItemSchema;

add_action('init', function() {
	$extend_instance = Package::container()->get( ExtendRestApi::class );

$extend_instance->register_endpoint_data(
		CartItemSchema::IDENTIFIER,
		"subscriptions",
		function () {
			return [
				'key' => [
					'description' => 'Subscription key',
					'type' => 'string',
				]
			];
		},
		function ( $cart_item ) {
			return [
				"key" => $cart_item['key']
			];
		},
	);
});
```

### How to test this change:

- Add the code example above in any plugin (even the main file of this plugin).
- In cart, inspect `store/cart` response.
- Under `items` you should see `extensions` and inside it `subscriptions` containing the data you added.
- Modify the code so that it causes an error, instead of `'key' => $cart_item['key']` try `'key' => new SomeClass`
- Visit cart again, `extensions` shouldn't include `subscriptions` anymore, an error would be logged.
- If you have debug enabled and logged in as an admin, you should an error being thrown.
